Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -0,0 +1,341 @@
+    tl.debug("win: " + win);
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug("Candidates found for match: " + allFiles.length);
+    for (let i = 0; i < allFiles.length; i++) {
+        tl.debug("file: " + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    let matchingFilesSet: Set<string> = new Set();
+
+    for (let i = 0; i < ftpOptions.filePatterns.length; i++) {
+        let normalizedPattern: string = path.join(
+            ftpOptions.rootFolder,
+            path.normalize(ftpOptions.filePatterns[i])
+        );
+
+        tl.debug("searching for files, pattern: " + normalizedPattern);
+
+        let matched = tl.match(allFiles, normalizedPattern, null, matchOptions);
+        tl.debug("Found total matches: " + matched.length);
+        // ensure each result is only added once
+        for (let j = 0; j < matched.length; j++) {
+            let match = path.normalize(matched[j]);
+            let stats = tl.stats(match);
+            if (!ftpOptions.preservePaths && stats.isDirectory()) {
+                // if not preserving paths, skip all directories
+            } else if (matchingFilesSet.add(match)) {
+                tl.debug(
+                    "adding " + (stats.isFile() ? "file:   " : "folder: ") + match
+                );
+                if (stats.isFile() && ftpOptions.preservePaths) {
+                    // if preservePaths, make sure the parent directory is also included
+                    let parent = path.normalize(path.dirname(match));
+                    if (matchingFilesSet.add(parent)) {
+                        tl.debug("adding folder: " + parent);
+                    }
+                }
+            }
+        }
+    }
+    return Array.from(matchingFilesSet).sort();
+}
+
+async function run() {
+    tl.setResourcePath(path.join(__dirname, "task.json"));
+
+    const tries = 3;
+    const ftpOptions: FtpOptions = getFtpOptions();
+
+    if (!ftpOptions.serverEndpointUrl.protocol) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FTPNoProtocolSpecified"));
+    }
+    if (!ftpOptions.serverEndpointUrl.hostname) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FTPNoHostSpecified"));
+    }
+
+    const files: string[] = findFiles(ftpOptions);
+    let ftpClient = await getFtpClient(ftpOptions);
+    const tracker = new ProgressTracking(ftpOptions, files.length + 1); // add one extra for the root directory
+
+    let retryWithNewClient = async (task: () => {}, retry: number) => {
+        let e = null;
+        while (retry > 0) {
+            try {
+                await task();
+                return;
+            } catch (err) {
+                e = err;
+                tl.warning(err);
+                ftpClient.close();
+
+                await sleep(1000);
+                ftpClient = await getFtpClient(ftpOptions);
+
+                retry--;
+            }
+        }
+
+        if (e) {
+            throw e;
```

Comment:
I think this function should just return an array of errors. The caller can decide what to do with them. But rethrowing will lose the original stack trace.

Reply:
This is interesting, I did some research and they claim losing original stack trace is a bug for chrome and looks to be fixed? 
https://stackoverflow.com/a/4434389
https://bugs.chromium.org/p/chromium/issues/detail?id=60240

Isn't node using chrome's javascript engine?

Reply:
Oh ok, that's good to know.

Still, I think the more modular choice makes for the better design, rather than just throwing the last error.

Reply:
Yeah, I definitely agree in principle.  It's just those tasks are tiny, and the error handling code is unlikely to be different.  If we make the failure handler function, it's yet another argument we have to pass in to a local method.